File: DistributedComputingAndSystems/Project/code-python/gameState.py
from gameStatus import GAME_STATUS
import random

from tools import gamePrint

test = False
testic = {
    'Dora': {'a': 0, 'b': 0, 'c': 28, 'd': 0}, 
    'Hari': {'a': 11, 'b': 14, 'c': 0, 'd': 0}, 
    'Filip': {'a': 0, 'b': 0, 'c': 0, 'd': 0}
    }

class GameState():
    def __init__(self) -> None:
        self.dic_TEST_ONLY = 0
        self.game_state = GAME_STATUS.NO
        self.leader = None
        self.players = []
        self.playersIndex = None
        self.onMovePlayer = None
        self.ps = None
    
    def setDicTEST_ONLY(self, num):
        self.dic_TEST_ONLY = num

    def setPLayersIndex(self):
        self.playersIndex = {}
        i = 0
        for p in self.players:
            self.playersIndex[p] = i
            i+= 1

    def startGame(self):
        if test:
            self.ps = testic
        else:
            self.ps = self.startPosition()
        self.setPLayersIndex()

    def getNextPlayer(self, player):
        return self.players[(self.players.index(player) +1) % len(self.players)]

    def relativeToApsolute(self, relative, playerIndex):
        if relative == 0:
            apsolute = 0
        elif relative <= 40:
            apsolute = (playerIndex * 10 + relative - 1) % 40 + 1
        else:
            apsolute = relative +  playerIndex * 10 
        return apsolute  

    def startPosition(self):
        dic = dict()
        for i in self.players:
            dic[i] = {'a':0,'b':0,'c':0,'d':0}
        return dic

    def play(self, player, forTestONLY = 0): #return to none
        die = random.randint(1,6) if self.dic_TEST_ONLY == 0 else self.dic_TEST_ONLY
        gamePrint("Your number on die is: "+ str(die))
        
        pieces  = self.getAvailablePieces(player,die)

        if len(pieces) == 0:
            gamePrint("You can not move any of your pieces")
            return (False,None)
        else:
            gamePrint("You can move on of following piceses: "+ str(pieces))
            move_piece = self.getMovePiece(pieces) if forTestONLY == 0 else forTestONLY
            if move_piece == "exit": return "exit"
            if self.ps[player][move_piece] == 0 and die == 6:
                self.ps[player][move_piece] = 1
            else:
                self.ps[player][move_piece] += die
                self.checkOthers(self.ps[player][move_piece], player) # if someone is killed
            cont = die == 6
            return (cont ,[move_piece,self.ps[player][move_piece]])
    
    def loadMove(self, player, move): 
        if move == None:
            return
        piece = move[0]
        position = move[1]
        self.ps[player][piece] = position
        self.checkOthers(self.ps[player][piece], player)

    def getWinner(self):
        for p in self.players:
            positions = list(self.ps[p].values())
            positions.sort()
            if [41,42,43,44] == positions:
                return p
        return None
      
    def getMovePiece(self,pieces):
        inp = input("\n")
        while inp not in pieces and inp != "exit":
            gamePrint("Enter one of available pieces")
            inp = input("\n")
        return inp

    def getAvailablePieces(self, player,die):
        items = self.ps[player].items()
        pieces = [k for k,v in items if v != 0 and v + die <= 40]
        taken_in_hause = [v for k, v in items if v > 40]
        pieces.extend([k for k,v in items if (v + die) > 40 and (v + die) <= 44 and
         (v + die) not in taken_in_hause])
        if die == 6: pieces.extend([k for k,v in items if v == 0])
        return sorted(pieces)

    def checkOthers(self,relative, player):
        absolute = self.relativeToApsolute(relative, self.playersIndex[player])
        for k, v in self.ps.items():
            if k == player:
                continue
            for piece, rel in v.items():
                if rel == 0:
                    continue
                piece_abs = self.relativeToApsolute(rel, self.playersIndex[k])
                if absolute == piece_abs:
                    self.ps[k][piece] = 0

    def stateToText(self):
        info = [self.leader, str(self.players), str(self.playersIndex),str(self.onMovePlayer), str(self.ps)]
        return "##".join(info)

    def setState(self, text):
        info = text.split("##")
        self.leader = info[0]
        self.players = eval(info[1])
        self.playersIndex = eval(info[2])
        self.onMovePlayer = eval(info[3])
        self.ps = eval(info[4])
        self.game_state = GAME_STATUS.YES

File: DistributedComputingAndSystems/Project/DSC_Project_java/src/wut/weiti/edcs/ludo/GameState.java
package wut.weiti.edcs.ludo;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;
import java.util.Scanner;

public class GameState {
	protected GameStatus gameStatus;
	protected String leader;
	protected String onMovePlayer;
	protected List<String> players;
	protected Map<String, Integer> playersIndex;
	protected Map<String,Map<String, Integer>> ps;
	protected int die_TEST_ONLY = 0;
	protected String piece_TEST_ONLY = null;
	
	public GameState() {
		gameStatus = GameStatus.NO;
		players = new ArrayList<String>();
		
	}
	
	public void setDie_TEST_ONLY(int die_TEST_ONLY) {
		this.die_TEST_ONLY = die_TEST_ONLY;
	}

	protected void startGame() {
		setPLayersIndex();
		startPosition();
	}
	protected void setPLayersIndex() {
		playersIndex = new HashMap<String, Integer>();
		int i = 0;
		for(String p :players) {
			playersIndex.put(p, i);
			i++;
		}
	}
	
	protected void startPosition() {
		ps = new HashMap<String, Map<String,Integer>>();
		for(String p :players) {
			Map<String, Integer> tmp = new HashMap<String, Integer>();
			tmp.put("a", 0);
			tmp.put("b", 0);
			tmp.put("c", 0);
			tmp.put("d", 0);
			ps.put(p, tmp);
			
		}
	}
	
	public String getNextPlayer(String player) {
		return this.players.get((this.players.indexOf(player) + 1) % this.players.size());
	}
	
	public int relativeToApsolute(int relative, int playerIndex) {
		if (relative == 0) return 0;
		if (relative <= 40) return (playerIndex * 10 + relative - 1) % 40 + 1;
		return relative + playerIndex * 10;
	}
	
	public Move play(String player) {
		int die  = die_TEST_ONLY == 0 ? ThreadLocalRandom.current().nextInt(1, 6 + 1) : die_TEST_ONLY;
		Tools.gamePrint("Your number on die is: "+ die);
		
		String[] pieces = getAvailablePieces(player, die);
		
		if (pieces.length == 0) {
			Tools.gamePrint("You can not move any of your pieces");
			return new Move(player, false, null, 0);
		}
		Tools.gamePrint("You can move on of following piceses: "+ Arrays.toString(pieces));
		String movePiece = piece_TEST_ONLY == null ? getMovePiece(pieces) : piece_TEST_ONLY;
		if (this.ps.get(player).get(movePiece) == 0 && die == 6)
			this.ps.get(player).put(movePiece, 1);
		else {
			Integer loc = this.ps.get(player).get(movePiece);
			this.ps.get(player).put(movePiece, loc + die);
			checkOthers(this.ps.get(player).get(movePiece), player);
		}
		boolean cont = die == 6;
		return new Move(player, cont, movePiece, this.ps.get(player).get(movePiece));
			
	}
	
	public void loadMove(Move move) {
		if (move.piece == null) return;
		this.ps.get(move.player).put(move.piece, move.position);
		checkOthers(this.ps.get(move.player).get(move.piece), move.player);
	}
	
	public String getWinner() {
		List<Integer> tmp = Arrays.asList(41,42,43,44);
		for (var p : this.players) {
			if (this.ps.get(p).values().containsAll(tmp)) return p;
		}
		return null;
	}

	protected void checkOthers(Integer relative, String player) {
		int absolute = this.relativeToApsolute(relative, this.playersIndex.get(player));
		this.ps.forEach((k,v) -> {
			if (k.equals(player)) return;
			v.forEach((piece, rel) -> {
				if (rel == 0) return;
				int piece_abs = relativeToApsolute(rel, playersIndex.get(k));
				if (absolute == piece_abs) {
					ps.get(k).put(k, 0);
				}
			});
		});
		
	}

	protected String getMovePiece(String[] pieces) {
		var list = Arrays.asList(pieces);
		Scanner scanner = new Scanner(System.in);
		while (scanner.hasNextLine()) {
			System.out.println();
			String input = scanner.nextLine();
			if (input.equals("exit")) throw new ExitException();
			if (list.contains(input)) return input;
			Tools.gamePrint("Enter one of available pieces");
		}
		return null;
	}

	protected String[] getAvailablePieces(String player, int die) {
		var psPlayer = this.ps.get(player).entrySet();
		List<Integer> takenInHause = psPlayer.
				stream().
				filter((e) ->  e.getValue() > 40).
				map(e -> e.getValue()).
				collect(Collectors.toList());
		return psPlayer.
				stream().
				filter((e) -> 
					(e.getValue() == 0 && die == 6 ) ||
					(e.getValue() + die <= 40 && e.getValue() != 0) ||
					(e.getValue() + die > 40 &&  e.getValue() + die <= 44 && !takenInHause.contains(e.getValue() + die))
				).
				map(e -> e.getKey()).
				collect(Collectors.toList()).toArray(new String[] {});
	}
	public String stateToText() {
		return this.leader + "##"+
				Tools.str2(players) +"##"+
				Tools.str1(playersIndex) + "##"+
				this.onMovePlayer == null ? "None": onMovePlayer + "##" +
				Tools.str(this.ps);
				
	}
	
	public void setState(String text) {
		String[] split = text.split("##");
		this.leader = split[0];
		this.players = Tools.eval2(split[1]);
		this.playersIndex = Tools.eval1(split[2]);
		this.onMovePlayer = split[3];
		this.ps = Tools.eval(split[4]);
		this.gameStatus = GameStatus.YES;
		
	}
	
}
